server/src/main/java/org/elasticsearch/cluster/metadata/MetaData.java

```diff
@@ -1371,6 +1372,25 @@ public MetaData build() {
             return aliasAndIndexLookup;
         }
 
+        private void validateDataStreams(SortedMap<String, AliasOrIndex> aliasAndIndexLookup) {
+            DataStreamMetadata dsMetadata = (DataStreamMetadata) customs.get(DataStreamMetadata.TYPE);
+            if (dsMetadata != null) {
+                for (DataStream ds : dsMetadata.dataStreams().values()) {
+                    if (aliasAndIndexLookup.containsKey(ds.getName())) {
+                        throw new IllegalStateException("data stream [" + ds.getName() + "] conflicts with existing index or alias");
+                    }
+
+                    final String backingIndexPrefixFrom = (ds.getName().startsWith(".") ? "" : ".") + ds.getName() + "-";
+                    final String backingIndexPrefixTo = (ds.getName().startsWith(".") ? "" : ".") + ds.getName() + ".";
+                    SortedMap<?, ?> map = aliasAndIndexLookup.subMap(backingIndexPrefixFrom, backingIndexPrefixTo);
+                    if (map.size() != 0) {
+                        throw new IllegalStateException(
+                            "data stream [" + ds.getName() + "] could create backing indices that conflict with existing indices");
```

Comment:
nit: a backing index could also conflict with an alias (or data stream in the future)